src/mapcss/__init__.py

```diff
@@ -113,44 +113,61 @@ def parseZoom(self, s):
+            self.choosers_by_type_zoom_tag[type] = {}
+        for zoom in range(self.minscale, self.maxscale + 1):
+            if zoom not in self.choosers_by_type_zoom_tag[type]:
+                self.choosers_by_type_zoom_tag[type][zoom] = {}
+            if clname not in self.choosers_by_type_zoom_tag[type][zoom]:
+                self.choosers_by_type_zoom_tag[type][zoom][clname] = {'arr': [], 'set': set()}
         if type in self.choosers_by_type:
             for chooser in self.choosers_by_type[type]:
-                for tag in chooser.extract_tags():
-                    if tag == "*" or tag == cltag:
-                        if chooser not in self.choosers_by_type_and_tag[type][clname]:
-                            self.choosers_by_type_and_tag[type][clname].add(chooser)
-                        break
-
-    def restore_choosers_order(self, type):
-        ethalon_choosers = self.choosers_by_type[type]
-        for tag, choosers_for_tag in list(self.choosers_by_type_and_tag[type].items()):
-            tmp = []
-            for ec in ethalon_choosers:
-                if ec in choosers_for_tag:
-                    tmp.append(ec)
-            self.choosers_by_type_and_tag[type][tag] = tmp
+                chooser_tags = chooser.extract_tags()
+                if '*' in chooser_tags or cltag in chooser_tags:
+                    for zoom in range(int(chooser.selzooms[0]), int(chooser.selzooms[1]) + 1):
+                        if chooser not in self.choosers_by_type_zoom_tag[type][zoom][clname]['set']:
+                            self.choosers_by_type_zoom_tag[type][zoom][clname]['arr'].append(chooser)
+                            self.choosers_by_type_zoom_tag[type][zoom][clname]['set'].add(chooser)
+
+    def finalize_choosers_tree(self):
+        for ftype in self.choosers_by_type_zoom_tag.keys():
+            for zoom in self.choosers_by_type_zoom_tag[ftype].keys():
+                for clname in self.choosers_by_type_zoom_tag[ftype][zoom].keys():
+                    # Discard unneeded unique set of choosers.
+                    self.choosers_by_type_zoom_tag[ftype][zoom][clname] = self.choosers_by_type_zoom_tag[ftype][zoom][clname]['arr']
+                    for i in range(0, len(self.choosers_by_type_zoom_tag[ftype][zoom][clname])):
+                        chooser = self.choosers_by_type_zoom_tag[ftype][zoom][clname][i]
+                        optimized = StyleChooser(chooser.scalepair)
+                        optimized.styles = chooser.styles
+                        optimized.eval_type = chooser.eval_type
+                        optimized.has_evals = chooser.has_evals
+                        optimized.has_runtime_conditions = chooser.has_runtime_conditions
+                        optimized.selzooms = [zoom, zoom]
+                        optimized.ruleChains = []
+                        for rule in chooser.ruleChains:
+                            # Discard chooser's rules that don't match type or zoom.
```

Comment:
Should we print warning here and above (unneeded)?
I suspect that we can find a lot of bugs in our current mapcss files ..

Reply:
Nope, those discarded rules are actually legit, we remove them just because we have pre-matched
		the choosers in this tree leaf to a specific (type, zoom, clname), so we discard rules that don't match for sure.
		But those rules will be used still in other tree leaves for other (type, zoom, clname) combinations.